### PR TITLE
Change automatic type inclusion rules

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Change `@types` lookup according to our discussion yesterday (cc @vladima @mhegazy )

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -972,7 +982,7 @@ namespace ts {
         return resolutions;
     }
 
-    export function getDefaultTypeDirectiveNames(options: CompilerOptions, rootFiles: string[], host: CompilerHost): string[] {
+    export function getDefaultTypeDirectiveNames(options: CompilerOptions, rootFiles: string[], host: TypeDirectiveResolutionHost): string[] {
```
Comment (vladima) [inline src/compiler/program.ts]: `/* @internal */`

File: src/compiler/program.ts [hunk 2]
```diff
@@ -772,6 +772,28 @@ namespace ts {
         mtime: Date;
     }
 
+    export function readDefaultTypeDirectiveNames(rootPath: string, sys: System): string[] {
```
Comment (mhegazy) [inline src/compiler/program.ts]: internal?

Comment (amanuel2) [inline src/compiler/program.ts]: export serves same purpose as export in C++?

File: src/compiler/program.ts [hunk 3]
```diff
@@ -772,6 +772,28 @@ namespace ts {
...
+        const localTypes = combinePaths(rootPath, "types");
+        let result: string[] = [];
+        if (sys.directoryExists(localTypes)) {
+            result = result.concat(sys.getDirectories(localTypes));
+        }
+
+        // Walk up the directory spine until we find the first node_modules folder
+        let nodeProbePath = normalizeSlashes(rootPath);
+        while (getDirectoryPath(nodeProbePath) !== nodeProbePath) {
+            if (sys.directoryExists(combinePaths(nodeProbePath, "node_modules"))) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: would checking once for `sys.directoryExists(nodeTypesPath)` be sufficient?

File: src/compiler/program.ts (hunk 2 above)
Comment (mhegazy) [inline src/compiler/program.ts]: do not think you need `System` here. just `{ directoryExists(path: string): boolean; getDirectories(path: string): string[];}`

**Pull request closed without merging**